include/jemalloc/internal/cache_bin.h

```diff
@@ -101,12 +106,7 @@ cache_bin_alloc_easy(cache_bin_t *bin, bool *success) {
 	 * cacheline).
 	 */
 	*success = true;
-	ret = *(bin->avail - bin->ncached);
-	bin->ncached--;
-
-	if (unlikely(bin->ncached < bin->low_water)) {
-		bin->low_water = bin->ncached;
-	}
+	ret = *(bin->avail - (bin->ncached + 1));
```

Comment:
Looks great!

We probably should also consider moving the load to the beginning of this function, so that we can avoid the "+1" plus make this load happen sooner. We would need to make sure avail always points to somewhere valid though.

Reply:
Right I think cache_bin_t would need to be one larger to avoid a potential invalid load, will leave this for another diff.